include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -313,49 +363,144 @@ namespace internal
+                        res[4] += matrix_0[i] * b;
+                        res[5] += matrix_1[i] * b;
+                        res[6] += matrix_2[i] * b;
+                        res[7] += matrix_3[i] * b;
+                      }
+
+                    if (n_components > 2)
+                      {
+                        const Number c = in2[stride_in * i];
+                        res[8] += matrix_0[i] * c;
+                        res[9] += matrix_1[i] * c;
+                        res[10] += matrix_2[i] * c;
+                        res[11] += matrix_3[i] * c;
+                      }
                   }
               }
             if (add)
               {
-                out[0] += res0;
-                out[stride_out] += res1;
-                out[2 * stride_out] += res2;
-                out[3 * stride_out] += res3;
+                out0[0] += res[0];
+                out0[stride_out] += res[1];
+                out0[2 * stride_out] += res[2];
+                out0[3 * stride_out] += res[3];
+                if (n_components > 1)
+                  {
+                    out1[0] += res[4];
+                    out1[stride_out] += res[5];
+                    out1[2 * stride_out] += res[6];
+                    out1[3 * stride_out] += res[7];
+                  }
+                if (n_components > 2)
+                  {
+                    out2[0] += res[8];
+                    out2[stride_out] += res[9];
+                    out2[2 * stride_out] += res[10];
+                    out2[3 * stride_out] += res[11];
+                  }
               }
             else
               {
-                out[0]              = res0;
-                out[stride_out]     = res1;
-                out[2 * stride_out] = res2;
-                out[3 * stride_out] = res3;
+                out0[0]              = res[0];
+                out0[stride_out]     = res[1];
+                out0[2 * stride_out] = res[2];
+                out0[3 * stride_out] = res[3];
+                if (n_components > 1)
+                  {
+                    out1[0]              = res[4];
+                    out1[stride_out]     = res[5];
+                    out1[2 * stride_out] = res[6];
+                    out1[3 * stride_out] = res[7];
+                  }
+                if (n_components > 2)
+                  {
+                    out2[0]              = res[8];
+                    out2[stride_out]     = res[9];
+                    out2[2 * stride_out] = res[10];
+                    out2[3 * stride_out] = res[11];
+                  }
               }
-            out += 4 * stride_out;
+            out0 += 4 * stride_out;
+            if (n_components > 1)
+              out1 += 4 * stride_out;
+            if (n_components > 2)
+              out2 += 4 * stride_out;
           }
         if (nn - nn_regular == 3)
           {
-            Number res0, res1, res2;
+            Number res0, res1, res2, res3, res4, res5, res6, res7, res8;
```

Comment:
why is this not an array? I would think you can shorten some of the code below by writing repeated statements as loops.
I think you can also only allocate necessary variables here.